src/shm/shm.c

```diff
@@ -0,0 +1,256 @@
+
+    switch (settings.role)
+    {
+    case SENDER:
+        return shared_sender(settings, fd, shm_info);
+        break;
+    case RECEIVER:
+        return shared_receiver(settings, fd, shm_info);
+        break;
+    default:
+        err_and_leave("Incorrect role", 5);
+        break;
+    }
+    return -1;
+}
+
+int shared_sender(setting_t settings, int lockfd, mmap_creat_ret_t mmap_info){
+    coord_struct* coord = (coord_struct*) mmap_info.mem_region;
+    if (coord->abort != 0){
+        shm_unlink(mmap_info.path);
+        return 1;
+    }
+    pthread_mutex_lock(&(coord->lock)); // Prevent any other thread from joining
+    pthread_mutex_lock(&(coord->writer_ready.lock));
+    if (coord->writer_ready.v != 0){
+        return 1;
+    }
+
+    mmap_creat_ret_t ret_copy_init = init_copy_area(settings.password, DEFAULT_WIDTH, &(coord->mmap_size));
+    copy_struct* copy = (copy_struct*) ret_copy_init.mem_region;
+    
+    // Lock the first active area, now it's ready to accept readers
+    pthread_mutex_lock(&(copy->active[0]));
+    coord->writer_ready.v = 1;
+
+    // Prevent the reader from signaling the writer
+    pthread_mutex_lock(&(copy->signal_wrtr.lock));
+
+    // At this point I'm holding: coord->lock, coord->writer_ready.lock, copy->active[0], copy->signal_wrtr.lock
+    pthread_mutex_unlock(&(coord->lock));
+
+    // Now check if readers have already arrived
+    pthread_mutex_lock(&(coord->reader_ready.lock));
+    if (coord->reader_ready.v != 0){
+        // A reader already arrived, just signal that writer is ready
+        pthread_mutex_unlock(&(coord->reader_ready.lock));
+
+        pthread_cond_signal(&(coord->writer_ready.cond)); // Have already set writer_ready.v
+        pthread_mutex_unlock(&(coord->writer_ready.lock));
+        // Here I hold: copy->active[0], copy->signal_wrtr.lock
+    } else {
+        // Have to wait for reader
+        pthread_mutex_unlock(&(coord->writer_ready.lock));
+        while (coord->reader_ready.v == 0){
+            pthread_cond_wait(&(coord->reader_ready.cond), &(coord->reader_ready.lock));
+        }   
+        // The writer will already have noticed that writer is ready
+        // No need to signal writer ready
+        pthread_mutex_unlock(&(coord->reader_ready.lock));
+        // Here I hold copy->active[0], copy->signal_wrtr.lock
+    }
+    coord->abort = 1; // Anyone else coming needs to know this area of memory ceased being useful
+
+    // Now I'm just holding the locks in copy, I need to subscribe to signal_wrtr until the reader arrived
+    // Holding copy->active[0] and copy->signal_wrtr.lock
+    copy->signal_wrtr.v = 0;
+    while (copy->signal_wrtr.v == 0){
+        pthread_cond_wait(&(copy->signal_wrtr.cond), &(copy->signal_wrtr.lock));
+    }
+    // I keep the lock on signal_wrtr, reader arrived and I can start writing
+    
+    int idx = 0;
+    FILE* fstr = fopen(settings.filename, "rb");
+    do
+    { // I have lock on active[idx] (and signal_wrtr.lock)
```

Comment:
Is not just active[active_segment] (about to be) locked (by reader) ?

Reply:
Duplicate, because the comment was suddenly missing in the »Files changed«-view.

Reply:
The order of execution (only relevant actions) is as follows:
1. Writer process locks active[0]
2. Writer process locks signal_wrtr
3. Writer process signals the reader that it is ready to start the copy process then waits on the signal writer condition
4. Readers accesses the copy_struct memory area
5. Reader locks leaving[1]
6. Reader signals the writer that it is ready to read
7. Reader tries to lock active[0]

Since the write already held the lock on active[0] before the reader could join the copy area (therefore preventing race conditions), this last step means that when the reader tries to lock active[0] it just blocks until the writer releases the lock on active[0] once it completed the writing process